ecommerce/api_test.py

```diff
@@ -357,6 +357,56 @@ def test_get_valid_coupon_versions_after_redemption(user, is_global):
     assert list(get_valid_coupon_versions(product, user)) == []
 
 
+@pytest.mark.parametrize("is_global", [True, False])
+def test_get_valid_coupon_versions_with_max_redemptions_per_user(user, is_global):
+    """
+    Verify that the correct CouponPaymentVersions are returned before and after redemption
+    """
+    AMOUNT = "1.00000"
+    COUPON_CODE = "TESTCOUPON1"
+    COUPON_TYPE = "promo"
+    MAX_REDEMPTIONS = 3  # Must be greater or equal to max_redemptions_per_user
+    MAX_REDEMPTIONS_PER_USER = 3
+
+    payment = CouponPaymentFactory()
+    coupon_payment_version = CouponPaymentVersionFactory(
+        payment=payment,
+        amount=Decimal(AMOUNT),
+        coupon_type=COUPON_TYPE,
+        max_redemptions=MAX_REDEMPTIONS,
+        max_redemptions_per_user=MAX_REDEMPTIONS_PER_USER,
+        num_coupon_codes=1,
+    )
+    coupon = CouponFactory(
+        payment=payment, coupon_code=COUPON_CODE, is_global=is_global
+    )
+    coupon_version = CouponVersionFactory(
+        payment_version=coupon_payment_version, coupon=coupon
+    )
+    expected_coupon_versions = [coupon_version]
+
+    products = [ProductFactory() for i in range(MAX_REDEMPTIONS_PER_USER + 1)]
+
+    if not is_global:
+        for product in products:
+            CouponEligibilityFactory.create(coupon=coupon, product=product)
+
+    for i in range(MAX_REDEMPTIONS_PER_USER + 1):
+        assert (
+            list(get_valid_coupon_versions(products[i], user, code=COUPON_CODE))
+            == expected_coupon_versions
+        )
+
+        if expected_coupon_versions:
+            redeem_coupon(
+                coupon_version=coupon_version,
+                order=OrderFactory.create(purchaser=user, status=Order.FULFILLED),
+            )
+
+        if i == MAX_REDEMPTIONS_PER_USER - 1:
+            expected_coupon_versions = []
```

Comment:
A few things with this test case:
1. Minor nitpick: for variables inside of a function I think it makes more sense to use lowercase
1. There are some optimizations that can be made to the way we use factories here
1. The logic was somewhat hard to follow with the several loops, conditionals, and variable reassignments. I edited some things in the suggestion below. I haven't actually run it but you could give it a try

```suggestion
    amount = "1.00000"
    coupon_code = "TESTCOUPON1"
    coupon_type = CouponPaymentVersion.PROMO
    max_redemptions = 3
    max_redemptions_per_user = 1
    
    coupon_payment_version = CouponPaymentVersionFactory.create(
        amount=Decimal(amount),
        coupon_type=coupon_type,
        max_redemptions=max_redemptions,
        max_redemptions_per_user=max_redemptions_per_user,
        num_coupon_codes=1,
    )
    coupon_version = CouponVersionFactory.create(
        payment_version=coupon_payment_version, 
        coupon__coupon_code=coupon_code, 
        coupon__ is_global=is_global,
    )
    products = ProductFactory.create_batch(max_redemptions_per_user + 1)

    if not is_global:
        CouponEligibilityFactory.create_batch(len(products), coupon=coupon, product=factory.Iterator(products))

    assert (
        list(get_valid_coupon_versions(products[0], user, code=coupon_code))
        == [coupon_version]
    )
    redeem_coupon(
        coupon_version=coupon_version,
        order=OrderFactory.create(purchaser=user, status=Order.FULFILLED),
    )
    # max_redemptions_per_user was set to 1, so the coupon should no longer be valid for this user
    assert (
        list(get_valid_coupon_versions(products[1], user, code=coupon_code))
        == []
    )
```

Reply:
Thanks @gsidebo  for the detailed review.

I have updated the code, but would like to mention that point 3 makes sense, loops and variable reassignments are hard to go through, but I think if we use the mentioned approach where `max_redemption_per_user` is equal to 1, then it loses its importance and will become a replica of already written tests (i.e: test_get_valid_coupon_versions_after_redemption).

The current test validates the following test-cases:
1) Ensures that a coupon can be used multiple times by increasing max_redemption_per_user value (as this is the primary feature added rn)
2) Ensures that the coupon can be used exactly the same amount of times which max_redemption_per_user defines.

To test these two scenarios, either we repeat the coupon redemption code or use a loop and variable reassignment. As the loop doesn't seem readable so I have refactored it as well.